File: #164_Maximum_Gap/Solution.java
/* Bucket sort */

import java.util.*;

public class Solution {
    public int maximumGap(int[] nums) {
        if (nums == null || nums.length < 2) return 0;
        int min = Integer.MAX_VALUE, max = Integer.MIN_VALUE;
        /* find min & max */
        for (int i = 0; i < nums.length; i++) {
        	min = Math.min(min, nums[i]);
        	max = Math.max(max, nums[i]);
        }
        /* bucket length */
        int blen = (max - min) / (nums.length - 1);
        if (blen == 0) blen += 1; // corner case
        // System.out.println("blen: " + blen);
        /* number of buckets */
        int n = (max - min) / blen + 1;
        // System.out.println("n: " + n);
        /* initial min & max for each buckets */
        int[][] buckets = new int[n][2];
        for (int i = 0; i < n; i++) {
        	buckets[i][0] = -1;
        	buckets[i][1] = -1;
        }
        /* divide nums into buckets */
        for (int i = 0; i < nums.length; i++) {
        	int index = (nums[i] - min)/ blen;
        	if (buckets[index][0] == -1) {
        		buckets[index][0] = nums[i];
        		buckets[index][1] = nums[i];
        	}
        	else {
        		buckets[index][0] = Math.min(buckets[index][0], nums[i]);
        		buckets[index][1] = Math.max(buckets[index][1], nums[i]);
        	}
        }
        /* find the max gap */
        int result = 0;
        int prevIndex = 0;
        for (int i = 1; i < n; i++) { // need to skip empty bucket
        	if (buckets[i][0] == -1) continue; 
        	result = Math.max(result, buckets[i][0] - buckets[prevIndex][1]);
        	prevIndex = i;
        }
        return result;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[] t = {1, 1, 1, 1, 1, 5, 5, 5, 5, 5};
    	System.out.println(sol.maximumGap(t));
    }
}

File: #164_Maximum_Gap/Solution.py
class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def maximumGap(self, nums):
        if not nums or len(nums) < 2:
        	return 0
        maxVal = max(nums)
        minVal = min(nums)
        diff = max((maxVal - minVal) / (len(nums) - 1), 1)
        minBucket = [-1 for x in range((maxVal - minVal) / diff + 1)]
        maxBucket = [-1 for x in range((maxVal - minVal) / diff + 1)]
        for n in nums:
        	ind = (n - minVal) / diff
        	if minBucket[ind] == -1:
        		minBucket[ind] = n
        	else:
        		minBucket[ind] = min(minBucket[ind], n)
        	if maxBucket[ind] == -1:
        		maxBucket[ind] = n
        	else:
        		maxBucket[ind] = max(maxBucket[ind], n)
        maxGap = 0
        x = 0
        prev = 0
        while x < len(minBucket):
        	if maxBucket[x] != -1:
        		prev = x
        		x += 1
        		break
        	x += 1
        while x < len(minBucket):
        	if minBucket[x] == -1:
        		x += 1
        		continue
        	maxGap = max(minBucket[x] - maxBucket[prev], maxGap)
        	prev = x
        	x += 1
        return maxGap

sol = Solution()
print sol.maximumGap([1,1,1,1,1,5,5,5,5,5])